src/tensors/cpu/ruy_adapter.h

```diff
@@ -0,0 +1,468 @@
+    }
+
+    float unquant_multiplier;
+  };
+};
+
+#endif
+
+/*
+ * The following nomenclature comes from intgemm. The current state of code is to keep the
+ * intgemm_interface.h diff minimal. There are possibly better abstractions.
+ */
+struct IntgemmViaRuy {
+  // Convert compile time errors into run-time ABORTS. This allows bringing in only int8_t and
+  // select functions that are required to create a path which will run while not achieving
+  // parity with intgemm.
+  template <class T>
+  struct IntBase {
+    using Type = T;
+    static void Quantize(const float *, Type *, float, Index) { ABORT("Quantize unsupported"); }
+
+    static void PrepareA(const float *input,
+                         Type *output,
+                         float quant_mult,
+                         Index rows,
+                         Index cols) {
+      ABORT("PrepareA Unsupported");
+    }
+
+    static void PrepareB(const float *, Type *, float, Index, Index) {
+      ABORT("PrepareB Unsupported");
+    }
+    static void PrepareBQuantizedTransposed(const Type *, Type *, Index, Index) {
+      ABORT("PrepareBQuantizedTransposed Unsupported");
+    }
+    static void PrepareBTransposed(const float *, Type *, float, Index, Index) {
+      ABORT("PrepareBTransposed Unsupported");
+    }
+    static void SelectColumnsB(const Type *, Type *, Index, const Index *, const Index *) {
+      ABORT("SelectColumnsB Unsupported");
+    }
+
+    template <class Callback>
+    static void Multiply(const Type *A_prepared,
+                         const Type *B_prepared,
+                         float *output,
+                         Index rows_A,
+                         Index width,
+                         Index cols_B,
+                         Callback callback) {
+      ABORT("Multiply (A*B) Unsupported");
+    }
+  };
+
+  // Intgemm nomenclature expects Int8. Missing functions are ABORTs.
+  struct Int8 : IntBase<int8_t> {
+    using Type = int8_t;
+    static void PrepareBQuantizedTransposed(const Type *input,
+                                            Type *output,
+                                            Index rows,
+                                            Index cols) {
+      std::memcpy(output, input, /*count=*/sizeof(Type) * (rows * cols));
+    }
+
+    static void PrepareBTransposed(const float *input,
+                                   Type *output,
+                                   float quant_mult,
+                                   Index rows,
+                                   Index cols) {
+      Preprocess<kHighestPath>::quantize(input, output, quant_mult, rows, cols);
+    }
+
+    static void PrepareA(const float *input,
+                         int8_t *output,
+                         float quant_mult,
+                         Index rows,
+                         Index cols) {
+      Preprocess<kHighestPath>::quantize(input, output, quant_mult, rows, cols);
+    }
+
+    static void SelectColumnsB(const Type *input,
```

Comment:
Marian's index_select can do the job of `selectColumnsB` if we do not do reorder B, so we should use that one instead. It is already used by default if not using intgemm: https://github.com/browsermt/marian-dev/blob/199201eb89b2941afdadb14164e936d412f897ad/src/layers/generic.cpp#L331
As long as the type is an 8bit type, index_select should just work.

Reply:
Unfortunately, no.

It is not clear to me at the moment how to connect this suggestion to an implementation. Within my limited understanding, the way to implement this suggestion is a different integration point in generic.cpp (replacing ifdefs alongside where WebAssembly gemm comes in):

https://github.com/browsermt/marian-dev/blob/199201eb89b2941afdadb14164e936d412f897ad/src/layers/generic.cpp#L296

based on a type-based if-else, creating an `ruygemm8` type, which we then configure to fall into the following branch https://github.com/browsermt/marian-dev/blob/199201eb89b2941afdadb14164e936d412f897ad/src/layers/generic.cpp#L331.  

I think this approach is significantly more effort and involves parts where my knowledge is limited - and I expect this will require touching a lot of other places and some refactoring to better abstract the incoming `ruygemm8` type. I could be wrong, but in either case, I don't have a path to realizing this suggestion without further help.